File: ios/Plugin/ZendeskSupport.swift
import Foundation
import SupportSDK
import ZendeskCoreSDK

@objc public class ZendeskSupport: NSObject {
    @objc public func initialize(_ appId: String,_ clientId: String,_ zendeskUrl: String,_ debugLog: Bool) throws {
        Zendesk.initialize(appId: appId, clientId: clientId, zendeskUrl: zendeskUrl)
        if (debugLog) {
            CoreLogger.enabled = debugLog
            CoreLogger.logLevel = .debug
        }
        Support.initialize(withZendesk: Zendesk.instance)
    }
    
    @objc public func setAnonymousIdentity(_ name: String,_ email: String) {
        let identity = Identity.createAnonymous(name: name, email: email)
        Zendesk.instance?.setIdentity(identity)
    }
    
    @objc public func setIdentity(_ token: String) {
        let identity = Identity.createJwt(token: token)
        Zendesk.instance?.setIdentity(identity)
    }
    
    @objc public func showHelpCenter(_ viewCtrl: UIViewController?,_ groupBy: String,_ groupIds: [NSNumber],_ labels: [String]) {
        let hcConfig = HelpCenterUiConfiguration()
        // Filter articles by group type
        if ("category" == groupBy && !groupIds.isEmpty) {
            hcConfig.groupType = .category
            hcConfig.groupIds = groupIds
        } else if ("section" == groupBy && !groupIds.isEmpty) {
            hcConfig.groupType = .section
            hcConfig.groupIds = groupIds
        }

        // Filter articles by label
        if (!labels.isEmpty) {
            hcConfig.labels = labels
        }
        
        DispatchQueue.main.async {
            let helpCenter = HelpCenterUi.buildHelpCenterOverviewUi(withConfigs: [hcConfig])
            let navController = UINavigationController(rootViewController: helpCenter)
            viewCtrl?.present(navController, animated: true, completion: nil)
            
        }
    }
    
    @objc public func showHelpCenterArticle(_ viewCtrl: UIViewController?, _ articleId: String) {
        DispatchQueue.main.async {
            let articleController = HelpCenterUi.buildHelpCenterArticleUi(withArticleId: articleId, andConfigs: [])
            let navController = UINavigationController(rootViewController: articleController)
            viewCtrl?.present(navController, animated: true, completion: nil)
        }
    }
    
    @objc public func showTicketRequest(_ viewCtrl: UIViewController?,_ subject: String,_ tags: [String],_ fields: [String]) {
            let requestConfig = RequestUiConfiguration()
        if (!subject.isEmpty) {
            requestConfig.subject = subject
        }
        if (!tags.isEmpty) {
            requestConfig.tags = tags
        }
        if (!fields.isEmpty) {
            var mappedFields: [CustomField] = []
            for field in fields {
                let fieldParts = field.components(separatedBy: "|")
                if (fieldParts.count == 2) {
                    let fieldId = (fieldParts[0] as NSString).longLongValue
                    let fieldValue = fieldParts[1]
                    
                    mappedFields.append(CustomField(fieldId: fieldId, value: fieldValue))
                }
            }
        }
        
        DispatchQueue.main.async {
            let requestController = RequestUi.buildRequestUi(with: [requestConfig])
            let navController = UINavigationController(rootViewController: requestController)
            viewCtrl?.present(navController, animated: true, completion: nil)
        }
    }
    
    @objc public func showUserTickets(_ viewCtrl: UIViewController?) {
        DispatchQueue.main.async {
            let requestListController = RequestUi.buildRequestList()
            let navController = UINavigationController(rootViewController: requestListController)
            viewCtrl?.present(navController, animated: true, completion: nil)
        }
    }
}

File: android/src/main/java/com/websquare/plugins/zendesksupport/ZendeskSupport.kt
package com.websquare.plugins.zendesksupport

import android.content.Context
import com.zendesk.logger.Logger
import zendesk.core.Zendesk
import zendesk.support.Support
import zendesk.core.AnonymousIdentity
import zendesk.core.Identity
import zendesk.core.JwtIdentity
import zendesk.support.guide.HelpCenterConfiguration
import zendesk.support.guide.HelpCenterActivity
import zendesk.support.guide.ViewArticleActivity
import zendesk.support.request.RequestConfiguration
import zendesk.support.request.RequestActivity
import zendesk.support.CustomField
import zendesk.support.requestlist.RequestListActivity
import java.util.ArrayList

class ZendeskSupport {
    // initialize zendesk support sdk
    fun initialize(context: Context?, url: String?, appId: String?, clientId: String?, debugLog: Boolean) {
        Zendesk.INSTANCE.init(context!!, url!!, appId!!, clientId)
        if (debugLog) Logger.setLoggable(true)
        Support.INSTANCE.init(Zendesk.INSTANCE)
    }

    fun setAnonymousIdentity(name: String?, email: String?) {
        val identity = AnonymousIdentity.Builder()
                .withNameIdentifier(name)
                .withEmailIdentifier(email)
                .build()
        Zendesk.INSTANCE.setIdentity(identity)
    }

    fun setIdentity(token: String?) {
        val identity: Identity = JwtIdentity(token)
        Zendesk.INSTANCE.setIdentity(identity)
    }

    fun showHelpCenter(context: Context?, groupBy: String, groupIds: List<Long?>, labels: List<String?>) {
//        Show Help Center Builder
        var builder = HelpCenterActivity.builder()
        if ("category" == groupBy && groupIds.isNotEmpty()) {
//            Filter articles by category
            builder = builder.withArticlesForCategoryIds(groupIds)
        } else if ("section" == groupBy && groupIds.isNotEmpty()) {
//            Filter articles by section
            builder = builder.withArticlesForSectionIds(groupIds)
        }
        //        Filter articles by label
        if (labels.isNotEmpty()) {
            builder = builder.withLabelNames(labels)
        }
        builder.show(context!!)
    }

    fun showHelpCenterArticle(context: Context?, articleId: String) {
//        Show a single article Builder
        ViewArticleActivity.builder(articleId.toLong()).show(context!!)
    }

    fun showTicketRequest(context: Context?, subject: String, tags: List<String?>, fields: List<String>) {
        var builder = RequestActivity.builder()
        if (subject.isNotEmpty()) {
            builder = builder.withRequestSubject(subject)
        }
        if (tags.isNotEmpty()) {
            builder = builder.withTags(tags)
        }
        if (fields.isNotEmpty()) {
            val mappedFields: ArrayList<CustomField?> = ArrayList()
            for (field in fields) {
                val fieldParts = field.split("\\|".toRegex()).toTypedArray()
                if (fieldParts.size == 2) {
                    val fieldId = fieldParts[0].toLong()
                    val fieldValue = fieldParts[1]
                    val customField = CustomField(fieldId, fieldValue)
                    mappedFields.add(customField)
                }
            }
            builder = builder.withCustomFields(mappedFields)
        }
        builder.show(context!!)
    }

    fun showUserTickets(context: Context?) {
        RequestListActivity.builder().show(context!!)
    }
}